### ISSUE TITLE
Ansi Color On Linux

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/corefx#29648]
_From @abk-x on May 10, 2018 4:10_

Looking for some guidance on how to debug this issue. How is dotnet cli determining if Ansi support is available? (Edit: I have seen https://github.com/dotnet/cli/issues/1977, but this was a very long time ago and I believe that this was resolved)

## Steps to reproduce
1. `dotnet new console`
2. `dotnet add package Microsoft.DotNet.Cli.Utils --version 2.0.0` (this is what dotnet build/run uses)
3. Following code:

```
using System;
using Microsoft.DotNet.Cli.Utils;

namespace test
{
    class Program
    {
        static void Main(string[] args)
        {
            var console = AnsiConsole.GetOutput();
            console.WriteLine("\x1b[31mRedText");

            Console.ForegroundColor = ConsoleColor.Red;
            Console.WriteLine("RedText2");
        }
    }
}
```
4. `dotnet run`

## Expected  behavior
Both RedText and RedText2 should appear red.

## Actual behavior
Both text appear white.
Additionally:

`$ script -q -c "dotnet run" out.txt && cat -A out.txt` produces:
```
Script started on 2018-05-10 13:53:40+1000$
^[[?1h^[=^[[?1h^[=^[[?1h^[=^[[39;49mRedText^M$
^[[39;49mRedText2^M$
$
Script done on 2018-05-10 13:53:42+1000$
```

Which doesn't appear to contain the correct ANSI escape chars for the correct colors (red is 31, 39 is default console color) , furthermore, dotnet build does not seem to output colors, unless I specify `/clp:ForceConsoleColor` (which is only possible for dotnet build, not dotnet run). I have also tried setting DOTNET_CLI_CONTEXT_ANSI_PASS_THRU=1 in the environment, but this does not change the result.

## Environment data

$TERM is xterm-256color
$SHELL is bash

```
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300-rc1-008673
 Commit:    f5e3ddbe73

Runtime Environment:
 OS Name:     debian
 OS Version:  
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.300-rc1-008673/

Host (useful for support):
  Version: 2.1.0-rc1
  Commit:  eb9bc92051

.NET Core SDKs installed:
  2.1.300-rc1-008673 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.0-rc1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

_Copied from original issue: dotnet/cli#9250_

**Comment by rainersigwald** [comment]
_From @livarcocc on May 10, 2018 16:59_

I am not 100% sure I understand the question, sorry. In any case, You can look at the Write method to see how we honor the colors: https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/AnsiConsole.cs#L100.

Also, I would not recommend using this library for this sort of thing. It was originally publish as a way to share APIs helpful in build project tools, which have since lost their utility with the move from project.json to msbuild.

We have also stopped publishing updates to this library and the last one we pushed was 2.0.0.

As for build and run, the output there is controlled directly by MSBuild.

**Comment by rainersigwald** [comment]
_From @abk-x on May 10, 2018 22:6_

Right, so lets remove that library from the discussion. The following:

```
Console.ForegroundColor = ConsoleColor.Red;
Console.WriteLine("RedText2");
```

Combined with `dotnet run`, does not produce colors at all, I am convinced it must be something to do with my environment, as I have other systems where this produces color, I am trying to work out what the cause could be :(

Edit: Just to confirm, only dotnet run and dotnet build suffer this issue, other systems (clang, gcc) show colors fine.
Edit2: Rereading your comment, it sounds like I might need to investigate the cause from the MSBuild front, happy to close this and move discussion there.

**Comment by rainersigwald** [comment]
Based on discussion in https://github.com/Microsoft/msbuild/issues/3296, this sounds an awful lot like `Console.ForegroundColor` isn't working on this machine.

@abk-x, can you elaborate a bit more on your distro? AFAIK colors have worked everywhere I tried.

**Comment by khellang** [comment]
> It looks like for whatever reason, the color isn't set at all.

Is that using the code you posted above?

```csharp
Console.ForegroundColor = ConsoleColor.Red;
Console.WriteLine("RedText2");
```

It's really weird if `Console.ForegroundColor = ConsoleColor.Red;` would result in `write(1, "\33[39;49m", 8)`.

**Comment by khellang** [comment]
Ah, it's probably due to the `ForegroundColor` setter calling `RefreshColors`:

https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/ConsolePal.Unix.cs#L140-L144

And `RefreshColors` calling `WriteResetColorString`:

https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/ConsolePal.Unix.cs#L499-L522

Sounds like the subsequent call to `WriteSetColorString` is failing somehow. I wonder if there's something weird going on with the terminfo database in this case?

**Comment by abk-x** [comment]
Hmm! Interesting! Is there anything I can run on my system to confirm this? What does netcore expect to see within terminfo db?

**Comment by abk-x** [comment]
Interesting! TERM=xterm dotnet build
gives me colors!

**Comment by khellang** [comment]
> What does netcore expect to see within terminfo db?

Since it was able to figure out the reset string, I'm pretty sure it found and read a terminfo database, but it couldn't find format strings for `SetAnsiForeground` and `SetAnsiBackground`.

The logic for locating your terminfo database is as follows, where `term` is the value of your `TERM` environment variable:

https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/TermInfo.cs#L184-L214

If it can't find one based on your `TERM`, it'll fall back to the following list:

https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/TermInfo.cs#L175-L179

> Is there anything I can run on my system to confirm this?

I guess you could try running `infocmp` and see what it gives you?

**Comment by abk-x** [comment]
infocmp gives me:
```
#	Reconstructed via infocmp from file: /lib/terminfo/x/xterm-256color
xterm-256color|xterm with 256 colors,
	am, bce, ccc, km, mc5i, mir, msgr, npc, xenl,
	colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
	acsc=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
	bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
	clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
	csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
	cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
	cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
	cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
	dl=\E[%p1%dM, dl1=\E[M, ech=\E[%p1%dX, ed=\E[J, el=\E[K,
	el1=\E[1K, flash=\E[?5h$<100/>\E[?5l, home=\E[H,
	hpa=\E[%i%p1%dG, ht=^I, hts=\EH, ich=\E[%p1%d@,
	il=\E[%p1%dL, il1=\E[L, ind=\n, indn=\E[%p1%dS,
	initc=\E]4;%p1%d;rgb\:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
	invis=\E[8m, is2=\E[!p\E[?3;4l\E[4l\E>, kDC=\E[3;2~,
	kEND=\E[1;2F, kHOM=\E[1;2H, kIC=\E[2;2~, kLFT=\E[1;2D,
	kNXT=\E[6;2~, kPRV=\E[5;2~, kRIT=\E[1;2C, kb2=\EOE, kbs=^?,
	kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
	kdch1=\E[3~, kend=\EOF, kent=\EOM, kf1=\EOP, kf10=\E[21~,
	kf11=\E[23~, kf12=\E[24~, kf13=\E[1;2P, kf14=\E[1;2Q,
	kf15=\E[1;2R, kf16=\E[1;2S, kf17=\E[15;2~, kf18=\E[17;2~,
	kf19=\E[18;2~, kf2=\EOQ, kf20=\E[19;2~, kf21=\E[20;2~,
	kf22=\E[21;2~, kf23=\E[23;2~, kf24=\E[24;2~,
	kf25=\E[1;5P, kf26=\E[1;5Q, kf27=\E[1;5R, kf28=\E[1;5S,
	kf29=\E[15;5~, kf3=\EOR, kf30=\E[17;5~, kf31=\E[18;5~,
	kf32=\E[19;5~, kf33=\E[20;5~, kf34=\E[21;5~,
	kf35=\E[23;5~, kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q,
	kf39=\E[1;6R, kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~,
	kf42=\E[17;6~, kf43=\E[18;6~, kf44=\E[19;6~,
	kf45=\E[20;6~, kf46=\E[21;6~, kf47=\E[23;6~,
	kf48=\E[24;6~, kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q,
	kf51=\E[1;3R, kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
	kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
	kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
	kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
	kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
	kind=\E[1;2B, kmous=\E[M, knp=\E[6~, kpp=\E[5~,
	kri=\E[1;2A, mc0=\E[i, mc4=\E[4i, mc5=\E[5i, meml=\El,
	memu=\Em, oc=\E]104\007, op=\E[39;49m, rc=\E8,
	rep=%p1%c\E[%p2%{1}%-%db, rev=\E[7m, ri=\EM,
	rin=\E[%p1%dT, ritm=\E[23m, rmacs=\E(B, rmam=\E[?7l,
	rmcup=\E[?1049l\E[23;0;0t, rmir=\E[4l, rmkx=\E[?1l\E>,
	rmm=\E[?1034l, rmso=\E[27m, rmul=\E[24m,
	rs1=\Ec\E]104\007, rs2=\E[!p\E[?3;4l\E[4l\E>, sc=\E7,
	setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
	setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
	sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p5%t;2%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
	sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
	smcup=\E[?1049h\E[22;0;0t, smir=\E[4h, smkx=\E[?1h\E=,
	smm=\E[?1034h, smso=\E[7m, smul=\E[4m, tbc=\E[3g,
	u6=\E[%i%d;%dR, u7=\E[6n, u8=\E[?%[;0123456789]c,
	u9=\E[c, vpa=\E[%i%p1%dd,
```

It is worth noting that I don't have $TERMINFO set, perhaps this is the cause.

**Comment by khellang** [comment]
> It is worth noting that I don't have $TERMINFO set, perhaps this is the cause.

I doubt it. It should still be able to find `/lib/terminfo/x/xterm-256color` based on the fallback locations listed above.

Now, the question is what's wrong with this terminfo database 🤔

It's worth noting that Debain Buster ships with ncurses 6.1, which has had issues in the past; https://github.com/dotnet/corefx/issues/26966.

I'm not really a Linux person, so I really have no idea what I'm talking about here 🤷‍♂️

**Comment by abk-x** [comment]
Well, you gave a good analysis! I will dig into it an see if I can find a solution

**Comment by abk-x** [comment]
`var ad = System.TermInfo.Database.ReadActiveDatabase()` returns the database.

`var e = new System.TerminalFormatStrings(ad);`

The following however, returns -1 https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/ConsolePal.Unix.cs#L794

Which get normalized to 0, then down here, we see only printing of colors if > 0:

https://github.com/dotnet/corefx/blob/48ce3254e28f2dae2d9a5efe7dd42cfab119d287/src/System.Console/src/System/ConsolePal.Unix.cs#L551

However, the colors field DOES exist in the terminfo db, afaik (its 0x100), It seems it could be a classic integer overflow issue, I printed the bytes around `int maxColors = db.GetNumber(TermInfo.WellKnownNumbers.MaxColors)` and I observed that pos+1 and pos are `255`, 

Unfortunately I am not familiar with the format to say what is going on.

**Comment by khellang** [comment]
Oh, sorry, yes. On the phone, so just copy-pasted your code 😂

**Comment by abk-x** [comment]
```
SetAnsiForeground [%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m
SetAnsiBackground [%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m
```

**Comment by stephentoub** [comment]
@abk-x, could you attach a copy of your `/lib/terminfo/x/xterm-256color` file?

**Comment by abk-x** [comment]
[xt.zip](https://github.com/dotnet/corefx/files/1994966/xt.zip)

**Comment by abk-x** [comment]
fwiw there seems to be a tool named tic which will take the output of infocmp (posted before) and output the compiled terminfo db. I have just gone thru `find /lib/terminfo/ | grep color` and it looks like anything with `256` in the name (e.g. screen-256color) causes the same issue

**Comment by stephentoub** [comment]
Thanks for sharing the file, @abk-x.  I see the problem.  The fix in https://github.com/dotnet/corefx/pull/28786 was unfortunately incomplete.  I'll post a fix. cc: @joperezr

**Comment by stephentoub** [comment]
Re-opening to track a fix for the release/2.1 branch...

**Comment by danmosemsft** [comment]
fixed in 2.1
